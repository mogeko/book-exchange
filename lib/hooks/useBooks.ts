import useSWRInfinite, { type SWRInfiniteResponse } from "swr/infinite";
import handleQuery, { type QueryParamType } from "@/lib/utils/queryTools";
import useSWR, { type SWRResponse } from "swr";
import { faker } from "@faker-js/faker";

function useBooks(queryParam: QueryParamProps = {}) {
  const query = handleQuery("/books", queryParam);
  const { data, error, ...otherRes }: SWRResponse<BookTypes[]> = useSWR(query);

  return {
    books: data,
    isLoading: !error && !data,
    isError: error,
    ...otherRes,
  };
}

export function useBook(id: string | number, queryParam: QueryParamProps = {}) {
  const query = handleQuery(`/books/${id}`, queryParam);
  const { data, error, ...otherRes }: SWRResponse<BookTypes> = useSWR(query);

  return {
    book: data,
    isLoading: !error && !data,
    isError: error,
    ...otherRes,
  };
}

export function useBooksInfinite(queryParam: QueryParamInfiniteProps) {
  const { offset = 0, ...other } = queryParam;
  const opts = { revalidateFirstPage: false };
  const { data, error, ...otherRes }: SWRInfiniteResponse<BookTypes[]> =
    useSWRInfinite((index, previous) => {
      if (previous && !previous.length) return null;
      return handleQuery("/books", {
        offset: index * other.limit! + offset!,
        ...other,
      });
    }, opts);

  return {
    data: data,
    isLoading: !error && !data,
    isError: error,
    ...otherRes,
  };
}

export const exampleData = {
  title: faker.word.noun(20),
  cover: faker.image.image(1280, 1910),
  description: faker.lorem.paragraph(10),
  published: faker.date.past().toISOString(),
  publisher: faker.company.companyName(),
  tags: faker.lorem.words(1),
  author: faker.name.firstName(),
  rates: faker.datatype.number(100),
  id: faker.datatype.number(100),
};

export type BookTypes = Partial<typeof exampleData>;

type QueryParamProps = QueryParamType<keyof BookTypes>;

type QueryParamInfiniteProps = {
  limit: number;
} & Omit<QueryParamProps, "limit">;

export default useBooks;
