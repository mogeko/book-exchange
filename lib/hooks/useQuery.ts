import handleQuery from "@/lib/utils/queryTools";
import useSWR, { type SWRConfiguration } from "swr";
import useSWRInfinite, { type SWRInfiniteConfiguration } from "swr/infinite";

function useQuery<T, E = any>(
  url: `/api/${string}` | null,
  param = {},
  opts: SWRConfiguration<T, E> = {}
) {
  const query = url ? handleQuery(url, param) : null;
  const { data, error, ...otherRes } = useSWR<T, E>(query, opts);

  return {
    data: data,
    isLoading: !error && !data,
    isError: error,
    ...otherRes,
  };
}

export function useQueryInfinite<T, E = any>(
  getKey: (index: number, previous: T | null) => string | null,
  opts: SWRInfiniteConfiguration<T, E> = {}
) {
  const { data, error, ...otherRes } = useSWRInfinite<T, E>(getKey, opts);

  return {
    data: data,
    isLoading: !error && !data,
    isError: error,
    ...otherRes,
  };
}

export type Opts<T = any, E = any> = SWRConfiguration<T, E>;
export type OptsInfinite<T = any, E = any> = SWRInfiniteConfiguration<T, E>;

export default useQuery;
