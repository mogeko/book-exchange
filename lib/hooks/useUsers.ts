import handleQuery from "@/lib/utils/queryTools";
import useSWR, { type SWRResponse } from "swr";
import { faker } from "@faker-js/faker";

function useUsers(queryParam: ParamProps = {}) {
  const query = handleQuery("/users", queryParam);
  const { data, error }: SWRResponse<UserType[]> = useSWR(query);
  return {
    users: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export function useUser(id: string | number, queryParam = {}) {
  const query = handleQuery(`/users/${id}`, queryParam);
  const { data, error }: SWRResponse<UserType> = useSWR(query);
  return {
    user: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export const exampleData = {
  avatar: faker.image.avatar(),
  username: faker.internet.userName(),
  description: faker.lorem.paragraph(10),
  email: faker.internet.email(),
  city: faker.address.city(),
  id: faker.datatype.number(10),
};

export type UserType = Partial<typeof exampleData>;

interface ParamProps {
  limit?: number;
  page?: number;
}

export default useUsers;
