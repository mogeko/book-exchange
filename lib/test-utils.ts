import { render, renderHook } from "@testing-library/react";
import type { RenderOptions, RenderHookOptions } from "@testing-library/react";
import WrapProvider from "@/providers/providers";
import { SWRConfig } from "swr";
import React from "react";

function renderWithProvider(ui: React.ReactElement, options?: RenderOptions) {
  return render(ui, { wrapper: WrapProviderForTest, ...options });
}

function renderHookWithProvider<Result, Props>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props>
) {
  return renderHook(render, { wrapper: WrapProviderForTest, ...options });
}

const WrapProviderForTest: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  return React.createElement(
    WrapProvider,
    null,
    React.createElement(
      SWRConfig,
      { value: { provider: () => new Map() } },
      children
    )
  );
};

export * from "@testing-library/react";
export { renderHookWithProvider as renderHook };
export { renderWithProvider as render };
