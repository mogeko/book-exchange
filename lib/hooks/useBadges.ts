import useQuery from "./useQuery";
import { type SWRConfiguration } from "swr";

function useBadge(id: string, opts?: SWRConfiguration<BadgesType>) {
  const { data, isError } = useQuery<BadgesType>(`/badges/${id}`, {}, opts);
  const handleZero = (badge: number) => (badge === 0 ? null : badge);

  if (isError || !data) return { menus: [], user: [] };
  return {
    menus: data.menus.map(handleZero),
    user: data.user.map(handleZero),
  };
}

export interface BadgesType {
  uid: `${number}`;
  menus: number[];
  user: number[];
}

export default useBadge;
