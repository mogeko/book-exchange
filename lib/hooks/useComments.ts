import useQuery, { type Opts } from "./useQuery";

function useComments(param: ParamProps = {}, opts?: Opts<CommentsType>) {
  return useQuery<CommentsType>("/comments", param, opts);
}

export function useComment(id?: string, opts?: Opts<CommentType>) {
  return useQuery<CommentType>(id ? `/comments/${id}` : null, {}, opts);
}

export type CommentsType = ({
  meta: {
    short_review: string;
    rates: number;
  };
  responds: SubCommentType[];
  belongs_to: `bk${number}`;
} & Omit<SubCommentType, "belongs_to">)[];

interface SubCommentType {
  id: `cm${number}`;
  author_meta: {
    uid: `${number}`;
    username: string;
    email: string;
    avatar: string;
  };
  meta: {
    likes: number;
    dislike: number;
    created_at: string;
    location: string;
  };
  belongs_to: `cm${number}`;
  msg: string;
  is_folded: boolean;
}

export interface CommentType {
  cm_id: `cm${number}`;
  msg: string;
}

interface ParamProps {
  uid?: `${number}`;
  bkid?: `bk${number}`;
  limit?: number;
  page?: number;
}

export default useComments;
