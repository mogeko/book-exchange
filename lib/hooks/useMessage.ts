import useQuery, { type Opts } from "./useQuery";

function useMessage(opts?: Opts<MessageType>) {
  return useQuery<MessageType>("/msg", {}, opts);
}

export type MessageType = Partial<{
  id: `msg${number}`;
  key: `${string}_${string}`;
}>[];

export default useMessage;
