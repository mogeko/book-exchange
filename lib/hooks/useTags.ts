import useQuery from "@/lib/hooks/useQuery";
import { type SWRConfiguration } from "swr";

function useTags(opts?: SWRConfiguration<TagsType>) {
  return useQuery<TagsType>("/tags", {}, opts);
}

export interface TagsType {
  [tagsGroup: string]: string[];
}

export default useTags;
