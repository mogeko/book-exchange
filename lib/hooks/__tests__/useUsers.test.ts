import "@testing-library/jest-dom";
import { renderHook, waitFor } from "@/lib/test-utils";
import { useUser } from "@/lib/hooks/useUsers";

describe("useUsers", () => {
  it("should be defined", () => {
    expect(useUser).toBeDefined();
  });

  it("should return data", async () => {
    const { result } = renderHook(() => useUser("1000"));

    expect(result.current.data).toBeUndefined();

    await waitFor(() => {
      expect(result.current.data).toBeDefined();
      expect(result.current.data?.uid).toBe("1000");
    });

    expect(result.current).toMatchSnapshot();
  });

  it("return data if uid is empty", async () => {
    const { result } = renderHook(() => useUser());

    expect(result.current.data).toBeUndefined();

    await waitFor(() => {
      expect(result.current.data).toBeUndefined();
    });
  });
});
