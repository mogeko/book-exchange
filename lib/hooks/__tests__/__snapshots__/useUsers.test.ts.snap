// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useUsers should return data 1`] = `
Object {
  "data": Object {
    "avatar": "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1123.jpg",
    "birthdate": "Tue, 29 Jun 1965 01:55:39 GMT",
    "city": "Lake Rociomouth",
    "description": "Ut facilis et dolorem. Omnis qui neque velit. Dolore similique accusantium officiis adipisci. Dolorum cum et totam eveniet culpa qui quibusdam ad facere. Rerum distinctio atque consequatur amet iure sequi voluptas dolorem. Suscipit non quam. Deserunt omnis sint similique cum aliquid est sunt odio nemo. Inventore est ut ut magnam ea dolores ipsum ea recusandae. Et facere cumque quae fuga dolor ipsa dolor. Unde sit cupiditate. Sint aspernatur ullam a est repellendus ducimus. Maxime id est dolor nobis fuga.",
    "email": "Camylle_Grady61@hotmail.com",
    "fullname": "Alessandra Mueller",
    "uid": "1000",
    "username": "Britney.Skiles",
  },
  "error": undefined,
  "isLoading": false,
  "isValidating": false,
  "mutate": [Function],
}
`;
